src/patch.dev.js

```diff
@@ -77,26 +77,43 @@ if (React.createElement.isPatchedByReactHotLoader) {
 // https://github.com/gaearon/react-hot-loader/issues/249
 function isReactRouterish(type) {
   return type && (
-    type.displayName === 'Route' ||
-    type.name === 'Route' || // In case Ryan and Michael embrace ES6 classes
-    type.displayName === 'IndexRoute' ||
-    type.name === 'IndexRoute'
+    type.displayName === 'Router' ||
+    type.name === 'Router' // In case Ryan and Michael embrace ES6 classes
   );
 }
 
-const createElement = React.createElement;
-function patchedCreateElement(type, props, ...args) {
+function forceUpdateRouteAndChildRouteComponents(route) {
+  // TODO: check whether it is possible to also handle the `getComponent` case here
+  if (route.component && typeof route.component === 'function') {
+    // Side effect 😱
+    // Force proxies to update since React Router ignores new props.
+    resolveType(route.component);
+  }
+
+  // Child routes will be in `props.children` when defining routes using react
+  // elements and in `childRoutes` when using plain routes
+  var childRoutes = [].concat(route.childRoutes, route.props && route.props.children).filter(Boolean);
+
+  for (var _len = childRoutes.length, _key = 0; _key < _len; _key++) {
+    var childRoute = childRoutes[_key];
+    forceUpdateRouteAndChildRouteComponents(childRoute);
+  }
+}
+
+var createElement = React.createElement;
+function patchedCreateElement(type, props) {
   // Ideally we want to teach React Router to receive children.
   // We're not in a perfect world, and a dirty workaround works for now.
   // https://github.com/reactjs/react-router/issues/2182
-  if (
-    isReactRouterish(type) &&
-    props &&
-    typeof props.component === 'function'
-  ) {
-    // Side effect 😱
-    // Force proxies to update since React Router ignores new props.
-    resolveType(props.component);
+  if (isReactRouterish(type) && props) {
+    // A router's routes may be given as children or through
+    // the `routes` prop.
+    var topRoutes = [].concat(props.routes, props.children).filter(Boolean);
```

Comment:
Let’s move this initial loop into `forceUpdateRouteAndChildRouteComponents` as well.